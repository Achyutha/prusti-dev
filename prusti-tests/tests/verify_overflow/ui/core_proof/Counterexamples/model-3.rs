// compile-flags: -Punsafe_core_proof=true -Pcounterexample=true

use prusti_contracts::*;

#[trusted]
struct VecWrapper<T> {
    values: Vec<T>,
}

#[model]
struct VecWrapper<#[concrete] i8> {
    values: Seq<i8>,
}

impl VecWrapper<i8> {
    #[trusted]
    #[requires(self.model().values.len() > Int::new(index))]
    #[ensures(self.model().values[Int::new(index)] == result)]
    fn lookup(&self, index: i64) -> i8 {
        self.values[index as usize]
    }
}


#[requires(v.model().values.len() == Int::new(4))]
//once -Pcheck_overflows works with sequences these lines can be removed 
#[requires(v.model().values[Int::new(0)] < 10 )] 
#[requires(v.model().values[Int::new(0)] > 0 )]
#[requires(v.model().values[Int::new(1)] < 10 )]
#[requires(v.model().values[Int::new(1)] > 0 )]
#[requires(v.model().values[Int::new(2)] < 10 )]
#[requires(v.model().values[Int::new(2)] > 0 )]
#[requires(v.model().values[Int::new(3)] < 10 )]
#[requires(v.model().values[Int::new(3)] > 0 )]
fn test1(v: &VecWrapper<i8>) {
    assert!(v.lookup(0) + v.lookup(1) + v.lookup(2) + v.lookup(3) == 15)
}

#[requires(v.model().values.len() == Int::new(1) )]
#[ensures(v.model().values[0] == 1)]
fn test2(v: VecWrapper<i8>) {}

fn main(){}