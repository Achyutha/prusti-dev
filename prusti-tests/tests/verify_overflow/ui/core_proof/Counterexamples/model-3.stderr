error: [Prusti: verification error] the asserted expression might not hold
  --> $DIR/model-3.rs:36:5
   |
36 |     assert!(v.lookup(0) + v.lookup(1) + v.lookup(2) + v.lookup(3) == 15)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the failing assertion is here
  --> $DIR/model-3.rs:36:5
   |
36 |     assert!(v.lookup(0) + v.lookup(1) + v.lookup(2) + v.lookup(3) == 15)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: counterexample for "v"
       value:   ref(VecWrapper_model {
            values: Seq(4, 5, 7, 3, ),
        })
  --> $DIR/model-3.rs:35:10
   |
35 | fn test1(v: &VecWrapper<i8>) {
   |          ^
   = note: this error originates in the macro `assert` (in Nightly builds, run with -Z macro-backtrace for more info)

error: [Prusti: verification error] postcondition might not hold.
  --> $DIR/model-3.rs:40:11
   |
40 | #[ensures(v.model().values[0] == 1)]
   |           ^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the error originates here
  --> $DIR/model-3.rs:41:1
   |
41 | fn test2(v: VecWrapper<i8>) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: counterexample for "v"
       value:   VecWrapper_model {
            values: Seq(3, ),
        }
  --> $DIR/model-3.rs:41:10
   |
41 | fn test2(v: VecWrapper<i8>) {}
   |          ^
note: counterexample for "result"
       value:   ()
  --> $DIR/model-3.rs:41:29
   |
41 | fn test2(v: VecWrapper<i8>) {}
   |                             ^^

error: aborting due to 2 previous errors

